Fix load-and-stream for tablets

raphaelsc opened the pull request:
It might happen that multiple tablets co-habit the same shard, so we want load-and-stream to jump into a new streaming session for every tablet, such that the receiver will have the data properly segregated. That's a similar treatment we gave to repair. Today, load-and-stream fails due to sstables spanning more than 1 tablet in the receiver. 

Synchronization with migration is done by taking replication map, so migrations cannot advance while streaming new data. A bug was fixed too, where data must be streamed to pending replicas too, to handle case where migration is ongoing and new data must reach both old and new replica set. A test was added stressing this synchronization path.

Another bug was fixed in sstable loading, which expected sharder to not be invalidated throughout the operation, but that breaks during migrations.

Fixes #17315.

avikivity commented on sstables_loader.cc:
```diff
@@ -286,11 +347,20 @@ future<> sstable_streamer::stream_sstable_mutations(const dht::partition_range&
+                                          _messaging, _db.local(), table_id, std::move(sstables));
 
-    co_await streamer.stream(primary_replica_only);
+    co_await streamer->stream(primary_replica_only);
```
What guarantees that the effective replication map lock is held around this?

raphaelsc replied:
Sure.

avikivity commented:
Looks good

raphaelsc commented on sstables/sstable_directory.cc:
```diff
@@ -67,7 +67,7 @@ sstable_directory::sstable_directory(replica::table& table,
         table.get_sstables_manager(),
         table.schema(),
-        table.get_effective_replication_map()->get_sharder(*table.schema()),
+        [&] { return std::ref(table.get_effective_replication_map()->get_sharder(*table.schema())); },
```
Currently it has this requirement of caller not using it after yielding (which is true in sstable usage), but that is indeed too strict. Can it be done in a follow-up?

raphaelsc replied:
> We have auto_refreshing_sharder, can it help here?

I think so. Will give it a try, thanks.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6856/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6856/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/37/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6856/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8077/)
#### Build Details:
- Duration: 2 hr 14 min
- Builder: spider8.cloudius-systems.com

raphaelsc commented on sstables_loader.cc:
```diff
@@ -136,7 +136,7 @@ class sstable_streamer {
         // it in descending order, to start from the smaller tokens.
         std::ranges::sort(_sstables, [] (const sstables::shared_sstable& x, const sstables::shared_sstable& y) {
-            return x->compare_by_first_key(*y) > 0;
+            return x->compare_by_first_key(*y) < 0;
```
done: https://github.com/scylladb/scylladb/pull/17454

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6919/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6919/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8118/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/81/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6919/)
#### Build Details:
- Duration: 2 hr 1 min
- Builder: spider4.cloudius-systems.com

raphaelsc commented:
since v2:
- rebased
- simplified [Fix online SSTable loading with concurrent tablet migration](https://github.com/scylladb/scylladb/pull/17449/commits/467289a64273265f893271181744cd4acf1be609) with dht::auto_refreshing_sharder

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7023/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7023/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/164/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8204/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7023/)
#### Failed Tests (2/23444):
- [test_tablet_load_and_stream](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7023/testReport/junit/%28root%29/test_tablets/test_tablet_load_and_stream) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_load_and_stream)
- [topology_experimental_raft.test_tablets.debug.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7023/testReport/junit/%28root%29/non-boost%20tests/topology_experimental_raft_test_tablets_debug_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_tablets.debug.5)
#### Build Details:
- Duration: 1 hr 50 min
- Builder: spider6.cloudius-systems.com

asias commented:
> > * How much of the fix is tablet related (and how much isn't and needs backporting?)
> 
> this one "sstables_loader: Stream to pending replicas if needed" for example became needed with migration happening in parallel to load-and-stream. I am thinking it might be relevant with vnodes too if we allow load-and-stream in parallel to topology changes, as pending replica might miss writes if leaving replica started streaming and missed newer writes that came through load-and-stream. Not sure if we support that in our administritive guide. But it's not mentioned in the docs (https://opensource.docs.scylladb.com/stable/operating-scylla/nodetool-commands/refresh.html). Maybe load-and-stream is supposed to be serialized with topology changes, but as you cannot predict it with tablets (due to migrations that have already started prior to start of op), the patch is required for correctness. @asias please comment.


Load and stream is designed to be used for cluster restore or offline cluster migration. During restore procedure, it is not recommended to do topology change.

> 
> all the rest is tablet related only.
> 
> > * How come we did not catch it earlier with load-and-stream tests? Were they not running with tablets?
> 
> that's correct. it's important that the test stresses concurrent migration and load-and-stream, and this series adds one for it.

asias commented on sstables/sstable_directory.hh:
```diff
@@ -189,6 +190,14 @@ private:
+    sstable_directory(sstables_manager& manager,
+          schema_ptr schema,
+          std::unique_ptr<dht::sharder> sharder_ptr,
+          const dht::sharder* sharder,
```
Why do we have two sharder pointers?  Who will pass the unique_ptr version and who will pass the plain pointer version? Can we unify them?

raphaelsc replied:
This is private interface, not pretty I know, they're mutually exclusive, will try to unify them using std::variant<...>, so that's made explicit in the interface.

asias commented on sstables_loader.cc:
```diff
@@ -213,39 +227,48 @@ future<> sstables_loader::load_and_stream(sstring ks_name, sstring cf_name,
+future<> sstables_loader::load_and_stream(sstring ks_name, sstring cf_name,
+        ::table_id table_id, std::vector<sstables::shared_sstable> sstables, bool primary_replica_only) {
+    // streamer guarantees topology stability, for correctness, by holding effective_replication_map
+    // throughout its lifetime.
```
I think it is safer to take the erm early in the outside layer, .i.e, where we get the ssables lists in sstables_loader::load_new_sstables.  This prevents we get the ssables, then the erm is updated.

raphaelsc replied:
sstable_directory takes erm during its lifetime too, so we have stability during loading. Taking erm early is not a guarantee migration is not ongoing. It's not a problem if erm is updated before streaming, as long as it won't happen during streaming (which is true as streamer takes erm too). which shard a sstable is assigned to (where we'll stream it from), is not a requirement for correctness.

asias commented on sstables_loader.cc:
```diff
@@ -139,25 +140,33 @@ class sstable_streamer {
 
-    future<> stream(bool primary_replica_only);
-private:
+    virtual ~sstable_streamer() {}
```
Let's add some words in the commit log for the motivation for the virtualization.

raphaelsc replied:
Ok.

asias commented on sstables_loader.cc:
```diff
@@ -217,6 +217,8 @@ future<> sstable_streamer::stream_sstable_mutations(const dht::partition_range&
                         current_targets.resize(1);
                     }
+                    auto pending = _erm->get_pending_endpoints(current_dk.token());
+                    std::move(pending.begin(), pending.end(), std::back_inserter(current_targets));
```
This will break the case where the primary_replica_only is true but pending node is not a primary replica after the topology change.

asias replied:
Perhaps it is better to serialize the load and stream with topology changes and tablet migrations of the table we are loading.

raphaelsc replied:
@asias I don't understand why primary-replica-only is employed in load-and-stream. with repair -pr, we know repair will run on each node, so entire cluster will be repaired. but load-and-stream is push, not pull.

asias replied:
The reason we have primary-replica-only flag is to reducing the amount of data we stream. Users could use primary-replica-only during load and stream , and run repair after it bring data to all replicas.

asias commented on sstables_loader.cc:
```
Even though taking erm blocks migration, it doesn't wait for
its completion.
```

It is not clear what completion you meant. I assume you meant the previous streaming or repair triggered by the node operations.


If we serialize load&stream with topology changes, there is no need to worry about the pending nodes.  This is helpful for both vnode and tablet tables.

raphaelsc replied:
I see repair doesn't handle it yet

`        // FIXME: we need to wait for current tablet movement and disable future tablet movement`

asias commented on sstables_loader.cc:
```diff
@@ -148,12 +148,74 @@ class sstable_streamer {
+private:
+    future<> stream_fully_contained_sstables(const dht::partition_range& pr, std::vector<sstables::shared_sstable> sstables,
+                                             bool primary_replica_only) {
+        // FIXME: fully contained sstables can be optimized.
```
Let's add a little bit more details for the optimization. E.g, we can stream the whole file with file stream in case we want to stream the whole sst file.

asias commented on test/topology_experimental_raft/test_tablets.py:
```diff
@@ -753,3 +757,75 @@ async def test_tablet_count_metric_per_shard(manager: ManagerClient):
+    cql = manager.get_cql()
+    # Creates multiple tablets in the same shard
+    await cql.run_async("CREATE KEYSPACE test WITH replication = {'class': 'NetworkTopologyStrategy', 'replication_factor': 1}" \
+                        " AND tablets = {'initial': 5};") # 5 is rounded up to next power-of-two
```
Why not use 8 directly? Are you testing the round up?

raphaelsc replied:
Kind of, but it's irrelevant to the test itself. But it doesn't hurt.

raphaelsc commented:
since v3:
- simplified private sstable_directory ctor, by using std::variant
- reworked handling of migration during streaming, now taking into account primary_replica_only flag.

asias commented on sstables_loader.cc:
```diff
@@ -115,75 +115,217 @@ class send_meta_data {
+        }
+        return tm.get_endpoint_for_host_id(tinfo->pending_replica.host);
+    });
+    if (pending_replica_endpoint) {
```
Consider:

Old replicas: n1, n2, n3
New replicas: n1, n2, n4

With primary_replica_only = true, with this patch, sstable_streamer::get_endpoints would return n1. tablet_sstable_streamer::get_endpoints would return n1 and n4.  This is still not respecting primary_replica_only. 
Since the new replica node n4 is not the primary replica node, even if it is a pending node, we should not stream to it. tablet_sstable_streamer::get_endpoints should return n1.

raphaelsc replied:
Actually it wouldn't. See tablet version uses base one for getting natural endpoints. So n1,n2 and n3 is trimmed to n1. Then we realize n4 is not replacing n1 but n3 instead. So we just return n1

raphaelsc replied:
I am not at computer now but you can verify what I said by looking at none_of... It checks if none of the current ranges is leaving (in the example with primary replica only set, it would be n1) and if so we won't include pending replica == n4.

asias replied:
You are right. The none_of check is making sure we do not return n4 in this case.
